### PR TITLE
Categorical encoders

### LINKED ISSUE #355: Add CategoricalEncoderTransforms
### 🚀 Feature Request

Adapt categorical encoders from Sklearn into etna

### Motivation

Really frustrating when such simple transform is absent in etna

### Proposal

Create sklearn converter for
https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.LabelEncoder.html
https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.LabelBinarizer.html

Create 2 classes LabelEncoderTransform and LabelBinarizerTransform
Should work with int and str types.

If meet unknown label:
1. LabelEncoderTransform should use mean label value
2. LabelBinarizerTransform should use mean label vector value

The return column should be categorical.

Inverse transform is not necessary (but it should be easy to do)

### Test cases

_No response_

### Alternatives

_No response_

### Additional context

_No response_

### Checklist

- [x] I discussed this issue with ETNA Team

### PR DESCRIPTION
**Pull request by Ama16**
**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

## Before submitting (must do checklist)

- [x] Did you read the [contribution guide](https://github.com/tinkoff-ai/etna/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs? We use Numpy format for all the methods and classes.
- [x] Did you write any new necessary tests?
- [x] Did you update the [CHANGELOG](https://github.com/tinkoff-ai/etna/blob/master/CHANGELOG.md)?

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Closing issues
Closes #355

### CODE REVIEW

File: etna/transforms/encoders/categorical.py [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+        ----------
+        in_column:
+            name of column to be resampled
+        inplace:
+            if True, apply resampling inplace to in_column, if False, add transformed column to dataset
+        out_column:
+            name of added column. If not given, use `self.__repr__()` or `regressor_{self.__repr__()}` if it is a regressor
+        strategy:
+            filling encoding in not fitted values:
+            - If "new_value", then replace missing dates with '-1'
```
Comment (Mr-Geekman) [inline etna/transforms/encoders/categorical.py:107]: It seems like "dates" isn't valid here.

File: etna/transforms/encoders/categorical.py [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+from sklearn.utils._encode import _check_unknown
+from sklearn.utils._encode import _encode
+from sklearn.utils._encode import _unique
+
+from etna.transforms.base import PerSegmentWrapper
+from etna.transforms.base import Transform
+
+
+class LE(preprocessing.LabelEncoder):
+    def transform(self, y: pd.Series, strategy):
```
Comment (Mr-Geekman) [inline etna/transforms/encoders/categorical.py:16]: It is better to create Enum class like in other cases like [this](https://github.com/tinkoff-ai/etna/blob/master/etna/transforms/missing_values/imputation.py)

File: etna/transforms/encoders/categorical.py (hunk 1 above)
Comment (iKintosh) [reply etna/transforms/encoders/categorical.py:107]: better to use `timestamps`

File: etna/transforms/encoders/categorical.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+from sklearn import preprocessing
+from sklearn.utils._encode import _check_unknown
+from sklearn.utils._encode import _encode
+from sklearn.utils._encode import _unique
+
+from etna.transforms.base import PerSegmentWrapper
+from etna.transforms.base import Transform
+
+
+class LE(preprocessing.LabelEncoder):
```
Comment (iKintosh) [inline etna/transforms/encoders/categorical.py:15]: Make it `_LabelEncoder`, because this class is not intended for users and longer name makes it easier to understand class meaning.

File: etna/transforms/encoders/categorical.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+            filling_value = np.mean(encoded[~np.isin(y, diff)])
+        else:
+            raise ValueError(f"There are no '{strategy}' strategy exists")
+
+        encoded[index] = filling_value
+        return encoded
+
+
+class _OneSegmentLabelEncoderTransform(Transform):
+    """Replace the values in the column with the Label encoding"""
```
Comment (iKintosh) [inline etna/transforms/encoders/categorical.py:37]: There is no `inverse transform` method on purpose?

File: etna/transforms/encoders/categorical.py (hunk 1 above)
Comment (iKintosh) [reply etna/transforms/encoders/categorical.py:107]: better to use `values` here

File: etna/transforms/encoders/categorical.py [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+            - If "mean", then replace missing dates using the mean in encoded column
+            - If "none", then replace missing dates with None
+        """
+        self.in_column = in_column
+        self.inplace = inplace
+        self.out_column = self._get_out_column(out_column)
+        self.strategy = strategy
+        super().__init__(transform=_OneSegmentLabelEncoderTransform(self.in_column, self.out_column, self.strategy))
+
+    def _get_out_column(self, out_column: Optional[str]) -> str:
```
Comment (iKintosh) [inline etna/transforms/encoders/categorical.py:117]: it should be named `get_column_name` as in other transforms

File: etna/transforms/encoders/categorical.py [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+        if self.inplace:
+            return self.in_column
+        if out_column:
+            return out_column
+        if self.in_column.startswith("regressor"):
+            return "regressor_" + self.__repr__()
+        return self.__repr__()
+
+
+###############################
```
Comment (iKintosh) [inline etna/transforms/encoders/categorical.py:130]: Do we need to review what's below this line?

**Review by iKintosh — CHANGES_REQUESTED** [review]
Mostly okay. But need some changes. 

Kind reminder: if you think that PR is not ready for merge you should make it a draft)

File: etna/transforms/encoders/categorical.py [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+        self.out_column = out_column
+        self.ohe = preprocessing.OneHotEncoder(handle_unknown="ignore", sparse=False)
+
+    def _get_column_name(self) -> str:
+        """Get the `out_column` depending on the transform's parameters."""
+
+        if self.out_column:
+            return self.out_column
+        if self.in_column.startswith("regressor"):
+            temp_transform = LabelBinarizerTransform(in_column=self.in_column, out_column=self.out_column)
```
Comment (Mr-Geekman) [inline etna/transforms/encoders/categorical.py:175]: Can't we use here just `self.__repr__`?

File: etna/transforms/encoders/categorical.py [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+        result dataframe
+        """
+        result_df = df.copy()
+        result_df[
+            [self._get_column_name() + "_" + str(i) for i in range(len(self.ohe.categories_[0]))]
+        ] = self.ohe.transform(np.array(df[self.in_column]).reshape(-1, 1))
+        return result_df
+
+
+class LabelBinarizerTransform(PerSegmentWrapper):
```
Comment (Mr-Geekman) [inline etna/transforms/encoders/categorical.py:214]: Add description to the class. What is its purpose.

File: tests/test_transforms/test_encoders/test_categorical_transform.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+from etna.datasets import TSDataset
+from etna.datasets import generate_periodic_df
+from etna.transforms.encoders.categorical import LabelBinarizerTransform
+from etna.transforms.encoders.categorical import LabelEncoderTransform
+
+
+@pytest.fixture
+def two_df_with_new_values():
+    df1 = TSDataset.to_dataset(generate_periodic_df(3, "2020-01-01", 10, 2, n_segments=2))
```
Comment (Mr-Geekman) [inline tests/test_transforms/test_encoders/test_categorical_transform.py:12]: Use named parameters here because it is not obvious what is the meaning of all these parameters.

File: etna/transforms/encoders/categorical.py [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+            name of column to apply transform to
+        out_column:
+            name of added column.
+        strategy:
+            filling encoding in not fitted values:
+            - If "new_value", then replace missing dates with '-1'
+            - If "mean", then replace missing dates using the mean in encoded column
+            - If "none", then replace missing dates with None
+        inplace:
+            if True, apply resampling inplace to in_column, if False, add transformed column to dataset
```
Comment (Mr-Geekman) [inline etna/transforms/encoders/categorical.py:63]: Resampling is irrelevant here

File: etna/transforms/encoders/categorical.py [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+            - If "new_value", then replace missing values with '-1'
+            - If "mean", then replace missing values using the mean in encoded column
+            - If "none", then replace missing values with None
+        """
+        self.in_column = in_column
+        self.inplace = inplace
+        self.strategy = strategy
+        self.out_column = out_column
+        super().__init__(
+            transform=_OneSegmentLabelEncoderTransform(self.in_column, self.out_column, self.strategy, self.inplace)
```
Comment (Mr-Geekman) [inline etna/transforms/encoders/categorical.py:147]: Use keyword parameters here like in_column=self.in_column.
This can be a reason for warning during creation with out_column=None.

File: etna/transforms/encoders/categorical.py [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+    def _get_column_name(self) -> str:
+        """Get the `out_column` depending on the transform's parameters."""
+        if self.inplace and self.out_column:
+            warnings.warn("Transformation will be applied inplace, out_column param will be ignored")
+        if self.inplace:
+            return self.in_column
+        if self.out_column:
+            return self.out_column
+        if self.in_column.startswith("regressor"):
+            temp_transform = LabelEncoderTransform(
```
[the remark this answers is not in the document]
Comment (Mr-Geekman) [reply etna/transforms/encoders/categorical.py:80]: What is about this?

File: etna/transforms/encoders/categorical.py [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+
+class _LabelEncoder(preprocessing.LabelEncoder):
+    def transform(self, y: pd.Series, strategy: str):
+        diff = _check_unknown(y, known_values=self.classes_)
+
+        index = np.where(np.isin(y, diff))[0]
+
+        encoded = _encode(y, uniques=self.classes_, check_unknown=False).astype(float)
+
+        if strategy == "None":
```
Comment (Mr-Geekman) [inline etna/transforms/encoders/categorical.py:31]: Threre is no such value as "None", only "none", may be better to check this using Enum type, not manually.

File: tests/test_transforms/test_encoders/test_categorical_transform.py [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+
+@pytest.fixture
+def df_for_categorical_encoding():
+    return TSDataset.to_dataset(
+        generate_periodic_df(periods=4, start_time="2020-01-01", scale=10, period=3, n_segments=2)
+    )
+
+
+def test_label_encoder(df_for_categorical_encoding):
+    """Test LabelEncoderTransform correct works."""
```
Comment (Mr-Geekman) [inline tests/test_transforms/test_encoders/test_categorical_transform.py:29]: I think, that better description is:
"Test that LabelEncoderTransform works correct"

File: tests/test_transforms/test_encoders/test_categorical_transform.py [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+
+@pytest.mark.parametrize(
+    "strategy, expected_values",
+    [
+        ("new_value", np.array([[0, 0], [1, -1], [-1, -1]])),
+        ("None", np.array([[0, 0], [1, np.nan], [np.nan, np.nan]])),
+        ("mean", np.array([[0, 0], [1, 0], [0.5, 0]])),
+    ],
+)
+def test_new_value_label(two_df_with_new_values, strategy, expected_values):
```
Comment (Mr-Geekman) [inline tests/test_transforms/test_encoders/test_categorical_transform.py:44]: label_encoder in the name of the test is omitted.

same issue with word order in description

File: tests/test_transforms/test_encoders/test_categorical_transform.py [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+
+
+@pytest.fixture
+def df_for_categorical_encoding():
+    return TSDataset.to_dataset(
+        generate_periodic_df(periods=4, start_time="2020-01-01", scale=10, period=3, n_segments=2)
+    )
+
+
+def test_label_encoder(df_for_categorical_encoding):
```
Comment (Mr-Geekman) [inline tests/test_transforms/test_encoders/test_categorical_transform.py:28]: May be better to name "test_label_encoder_simple"

File: tests/test_transforms/test_encoders/test_categorical_transform.py (hunk 14 above)
Comment (Mr-Geekman) [inline tests/test_transforms/test_encoders/test_categorical_transform.py:29]: It will be more correct to use
"Test that LabelEncoderTransform works correct in simple case."

File: etna/transforms/encoders/categorical.py [hunk 17]
```diff
@@ -0,0 +1,235 @@
...
+        result dataframe
+        """
+        result_df = df.copy()
+        result_df[
+            [self._get_column_name() + "_" + str(i) for i in range(len(self.ohe.categories_[0]))]
+        ] = self.ohe.transform(np.array(df[self.in_column]).reshape(-1, 1))
+        return result_df
+
+
+class OneHotEncoderTransform(PerSegmentWrapper):
```
Comment (Mr-Geekman) [inline etna/transforms/encoders/categorical.py:217]: Describe how it works with new values.

File: etna/transforms/encoders/categorical.py [hunk 18]
```diff
@@ -0,0 +1,229 @@
...
+        result_df[[self.out_column + "_" + str(i) for i in range(len(self.ohe.categories_[0]))]] = result_df[
+            [self.out_column + "_" + str(i) for i in range(len(self.ohe.categories_[0]))]
+        ].astype("category")
+        return result_df
+
+
+class OneHotEncoderTransform(PerSegmentWrapper):
+    """
+    Encode categorical feature as a one-hot numeric features.
+    If unknown category is encountered during transform, the resulting one-hot encoded columns for this feature will be all zeros.
```
Comment (Mr-Geekman) [inline etna/transforms/encoders/categorical.py:202]: It seems like we have to add empty line after the first line with short description: [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard).

File: tests/test_transforms/test_encoders/test_categorical_transform.py [hunk 19]
```diff
@@ -0,0 +1,182 @@
...
+    train_ts, test_ts = ts_for_ohe_sanity
+    ohe = OneHotEncoderTransform(in_column="regressor_0")
+    filt = FilterFeaturesTransform(exclude=["regressor_0"])
+    train_ts.fit_transform([ohe, filt])
+    model = LinearPerSegmentModel()
+    model.fit(train_ts)
+    future_ts = train_ts.make_future(10)
+    forecast_ts = model.forecast(future_ts)
+    r2 = R2()
+    assert 1 - r2(forecast_ts, test_ts)["segment_0"] < 1e-5
```
Comment (Mr-Geekman) [inline tests/test_transforms/test_encoders/test_categorical_transform.py:49]: In metric we expect `y_true` to be the first parameter.

File: tests/test_transforms/test_encoders/test_categorical_transform.py [hunk 20]
```diff
@@ -0,0 +1,182 @@
...
+        else:
+            return 3 + rng.normal(0, 0.01)
+
+    df_to_forecast["segment_0", "target"] = df_regressors["segment_0"]["regressor_0"][:100].apply(f)
+    ts = TSDataset(df=df_to_forecast, freq="D", df_exog=df_regressors)
+    train_ts, test_ts = ts.train_test_split(test_size=10)
+    return train_ts, test_ts
+
+
+def test_ohe_sanity(ts_for_ohe_sanity):
```
Comment (Mr-Geekman) [inline tests/test_transforms/test_encoders/test_categorical_transform.py:39]: I think we should
1. Add description to the test.
2. Move it to the very of the file (fixture should be moved to the end of fixtures for this file).

File: tests/test_transforms/test_encoders/test_categorical_transform.py [hunk 21]
```diff
@@ -0,0 +1,182 @@
...
+        if x == 5:
+            return 10 + rng.normal(0, 0.01)
+        elif x == 8:
+            return 15 + rng.normal(0, 0.01)
+        else:
+            return 3 + rng.normal(0, 0.01)
+
+    df_to_forecast["segment_0", "target"] = df_regressors["segment_0"]["regressor_0"][:100].apply(f)
+    ts = TSDataset(df=df_to_forecast, freq="D", df_exog=df_regressors)
+    train_ts, test_ts = ts.train_test_split(test_size=10)
```
Comment (Mr-Geekman) [inline tests/test_transforms/test_encoders/test_categorical_transform.py:35]: I think we should move train_test_split into the test, because they have the same parameter horizon=10.

File: tests/test_transforms/test_encoders/test_categorical_transform.py [hunk 22]
```diff
@@ -0,0 +1,182 @@
...
+from etna.metrics import R2
+from etna.models import LinearPerSegmentModel
+from etna.transforms import FilterFeaturesTransform
+from etna.transforms.encoders.categorical import LabelEncoderTransform
+from etna.transforms.encoders.categorical import OneHotEncoderTransform
+
+
+@pytest.fixture
+def ts_for_ohe_sanity():
+    df_to_forecast = generate_ar_df(100, start_time="2021-01-01", n_segments=1)
```
Comment (Mr-Geekman) [inline tests/test_transforms/test_encoders/test_categorical_transform.py:16]: Add periods parameter in cases like this.

File: etna/transforms/encoders/categorical.py [hunk 23]
```diff
@@ -0,0 +1,229 @@
...
+            - If "new_value", then replace missing values with '-1'
+            - If "mean", then replace missing values using the mean in encoded column
+            - If "none", then replace missing values with None
+        """
+        self.in_column = in_column
+        self.inplace = inplace
+        self.strategy = strategy
+        self.out_column = out_column
+        self.out_column = self._get_column_name()
+        super().__init__(transform=_OneSegmentLabelEncoderTransform(self.in_column, self.out_column, self.strategy))
```
Comment (Mr-Geekman) [inline etna/transforms/encoders/categorical.py:129]: Use keyword parameters to create object.

**Review by iKintosh — DISMISSED** [review]
Seems like everything is okay.

**Review by Mr-Geekman — APPROVED** [review]
